### ISSUE TITLE
Freeze by endless loops in Physics

### ISSUE THREAD

**Issue by meekee7** [OpenTrespasser/JurassicParkTrespasser#84]
Sometimes the game freezes because it is stuck in an endless loop in physics code.

## Steps to reproduce ##
1. Pick up a "storable" object like a pistol or keycard.
2. Store the object in your inventory.
3. Move so that you stand directly in front of a wall (doors work too)
4. Keep storing and retrieving the object in such a way that the hand or the object touch the wall.

Personally I use the pistols in the security entrance at the beginning of the Industrial Town level for testing this.

The bug only rarely occurs in `Debug` builds, but is easy to reproduce in `Release` builds.

## The cause ##
The problem is caused by two loops in physics code that are supposed to normalize vectors. The code is almost the same for both. Those loops are `Pelvis.cpp` lines 1162 to 1187 and `Xob_bc.cpp` lines 3738 to 3763.

The loop is supposed to end when a `float` variable `L` becomes smaller than a certain threshold. However, the comparison does not work anymore when `L` becomes `NaN`. This happens because it was calculated from the vector entries, which are all `NaN` as well. Current research has shown that they were already `NaN` before the loop. Why this happens needs to be investigated.

## Solution approaches ##
Ending the loop when `L` becomes `NaN` prevents the game from freezing. But the vector is still full of `NaNs` and new problems occur immediately:
 - The player cannot jump anymore
 - When moving, there are very frequent little jumps. Sometimes the player can barely move at all
 - The arm cannot be used anymore
 - In summary, the game is left in an unplayable state.

An alternative approach is to do the bailout described above, but also set the vector that was supposed to be normalized to a valid default value. That default value needs yet to be determined. After a first test with a quad-0.25 vector, various physics glitches with held items occur, but the game remains playable.

The best solution is to find out why the vector became `NaN` in the first place.

**Comment by ell1e** [comment]
I saw a similar bug in the modern [bullet physics](https://github.com/bulletphysics/bullet3) library:

For "bullet physics", it was caused by velocity not being clamped after collisions to some reasonable maximum. If you don't do that, you have a constant risk of producing NaN velocities by super small lightweight objects colliding with super big heavyweight ones, and the small ones getting fired off at ever more ludicrous speeds where then they cause infinity values at some point, which in followup math calculations then often cause not a number.

So my suggestion to make such NaN poisoning less likely, without having looked at the Trespasser physics code yet, would be to look at the collision code and to clamp velocities after collisions. For some reason almost no physics engine does that by default, even though it greatly improves stability even in modern games.

**Comment by ell1e** [comment]
(For what it's worth, a wall might count as a "big heavyweight" object, and an inventoryu item could be a "small lightweight" one. So I wouldn't be surprised at all if the cause here was the same, velocities ending up too close at the numerical limit after such a collision, then ranodm followup collisions pushing one of the numbers to infinity, and then followup operations causing NaN down the line.)
